packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol

```diff
@@ -357,23 +371,62 @@ contract BridgePool is Testable, BridgePoolInterface, ExpandedERC20 {
         emit RelayDisputed(depositHash, keccak256(ancillaryData));
     }
 
-    /************************************
-     *           View FUNCTIONS         *
-     ************************************/
+    /**
+     * @notice Synchronize any balance changes in this contract with the utilized & liquid reserves. This would be done
+     * at the conclusion of an L2->L1 token transfer via the canonical token bridge.
+     */
+    function sync() public {
+        // Check if the l1Token balance of the contract is greater than the liquidReserves. If it is then the bridging
+        // action from L2->L1 has concluded and the local accounting can be updated.
+        uint256 l1TokenBalance = l1Token.balanceOf(address(this));
+        if (l1TokenBalance > liquidReserves) {
+            // utilizedReserves can go to less than zero. This will happen if the accumulated fees exceeds the current
+            // outstanding utilization. In other words, if outstanding bridging transfers are 0 then utilizedReserves
+            // will equal the total LP fees accumulated over all time.
+            utilizedReserves -= int256(l1TokenBalance - liquidReserves);
+            liquidReserves = l1TokenBalance;
+        }
+    }
 
     /**
      * @notice Computes the exchange rate between LP tokens and L1Tokens. Used when adding/removing liquidity.
      */
-    function exchangeRateCurrent() public view returns (uint256) {
+    function exchangeRateCurrent() public returns (uint256) {
         if (totalSupply() == 0) return 1e18; //initial rate is 1 pre any mint action.
 
-        // Consider a naive rate implementation. This acts like a step function, increasing when funds hit L1 from the
-        // canonical bridge. TODO: update with a more elaborate technique that pays out gradually over the 1 week loan.
+        // First, update fee counters and local accounting of finalized transfers from L2->L1.
+        updateFeeCounters(0); // Accumulate all allocated fees from the last time this method was called.
+        sync(); // Fetch any balance changes due to token bridging finalization and factor them in.
+
+        // ExchangeRate := (liquidReserves+utilizedReserves+cumulativeLpFeesEarned-undistributedLpFees)/lpTokenSupply
```

Comment:
I'm still thinking on this implementation, but one thought that pops into my head when reading this is why do we need to track cumulative lp fees earned?

Could we not just add LP fees to the utilized reserves as they're paid out and allow the utilized reserves to go negative in the case where there are some fees yet to be paid out, but nothing in transit?

The way I'm thinking about it (and I could totally be missing something, as this is very nuanced), the cumulativeLpFeesEarned feels like it could be unnecessary complexity.

Thoughts?

Reply:
You cant just add them to the utilized reserves as they don't technically count as utilized reserves. if anything, they are `liquidReserves` that are in flight. I think what you propose would work numerically but it's overloading the notion of a `utalizedReserve` making it harder to reason about.

I'm going to try out what you propose to double-check it works as expected in any case.

Reply:
I tried changing the exchange rate equation to this:
```
FixedPoint.Unsigned memory numerator =
            FixedPoint.Unsigned(liquidReserves).sub(FixedPoint.Unsigned(undistributedLpFees));
        if (utilizedReserves > 0) numerator = numerator.add(FixedPoint.Unsigned(uint256(utilizedReserves)));
        else numerator = numerator.sub(FixedPoint.Unsigned(uint256(utilizedReserves * -1)));
        return numerator.div(FixedPoint.Unsigned(totalSupply())).rawValue;
```

while updating the `updateFeeCounters` method to this:
```
function updateFeeCounters(uint256 allocatedLpFees) internal {
        // Calculate the unallocatedAccumulatedFees from the last time the contract was called.
        uint256 unallocatedAccumulatedFees = getUnallocatedAccumulatedFees();

        // Decrement the undistributedLpFees by the amount of accumulated fees.
        undistributedLpFees = undistributedLpFees > unallocatedAccumulatedFees
            ? undistributedLpFees - unallocatedAccumulatedFees
            : 0;

        lastLpFeeUpdate = getCurrentTime();

        // If this method was called from settleRelay then it includes the exact amount of fees to be attributed to the
        // Liquidity providers. add this to the associated variables.
        if (allocatedLpFees > 0) {
            undistributedLpFees += allocatedLpFees;
            utilizedReserves += int256(allocatedLpFees);
        }
    }
```

This effectively amounts to adding in the `allocatedLpFees` to the utilized reserves. After doing this, I could remove the `cumulativeLpFeesEarned` in totality.

I commented out all tests that tested `cumulativeLpFeesEarned` and `utilizedReserves` (as the former is removed after this and the latter's definition has changed somewhat) and ran all the tests. Every virtual balance accounting test works exactly as before (which is expected).

This means we can opt for this and save one local variable and make a marginal improvement in gas costs. The question need to be: is this worth the overloading and potentially higher complexity in understanding to bake this into the `utalizedReserves`? It might be but am curious to hear others thoughts.

Reply:
I've updated the implementation to track what you recommended optimistically. let me know if this matches to your expected format.